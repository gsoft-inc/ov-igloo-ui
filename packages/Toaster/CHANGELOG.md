# @igloo-ui/toaster

## 0.1.0

### Minor Changes

- f241a98: Toaster component

  The Toaster component is a non-disruptive message that appears at the top of the interface to provide quick, at-a-glance feedback on the outcome of an action.

  ### Component API

  | name                                  | description                   |
  | ------------------------------------- | ----------------------------- |
  | `toaster.success(message, duration)`  | render a success notification |
  | `toaster.success(message, duration)`Â  | render a error notification   |
