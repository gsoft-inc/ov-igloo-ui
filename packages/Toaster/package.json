{
  "name": "@igloo-ui/toaster",
  "version": "2.1.1",
  "main": "dist/Toaster.js",
  "module": "dist/Toaster.js",
  "types": "dist/Toaster.d.ts",
  "styles": "dist/toaster.css",
  "publishConfig": {
    "access": "public"
  },
  "repository": "git+https://github.com/gsoft-inc/ov-igloo-ui.git",
  "author": "Officevibe Inc.",
  "contributors": [
    "Franck franck.gaudin@gsoft.com"
  ],
  "license": "Apache-2.0",
  "peerDependencies": {
    "react": ">= 16.8.6",
    "react-dom": "^17.0.2"
  },
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "rollup -c rollup.config.js"
  },
  "dependencies": {
    "@igloo-ui/icon-button": "^1.2.1",
    "@igloo-ui/icons": "^1.12.2",
    "@igloo-ui/tokens": "^2.0.0",
    "@hopper-ui/tokens": "^3.1.4",
    "@react-aria/interactions": "^3.21.1",
    "@react-aria/toast": "^3.0.0-beta.7",
    "@react-stately/toast": "^3.0.0-beta.1",
    "classnames": "^2.3.2",
    "use-sync-external-store": "^1.2.0"
  },
  "devDependencies": {
    "@igloo-ui/button": "^0.8.0",
    "@types/react-dom": "^18.0.11",
    "@types/use-sync-external-store": "^0.0.3"
  },
  "browserslist": [
    "> 1%",
    "last 2 versions",
    "IE >= 11"
  ]
}
