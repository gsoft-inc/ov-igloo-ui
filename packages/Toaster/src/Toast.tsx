import * as React from "react";
import cx from "classnames";
import type { QueuedToast, ToastState } from "@react-stately/toast";
import { useToast } from "@react-aria/toast";

import IconButton from "@igloo-ui/icon-button";
import Close from "@igloo-ui/icons/dist/Close";
import SuccessSolid from "@igloo-ui/icons/dist/SuccessSolid";
import RemoveSolid from "@igloo-ui/icons/dist/RemoveSolid";
import type { ToastArgs } from "./Toaster";

export interface ToastProps<T> {
    /** The toast state used by react-aria */
    state: ToastState<T>;
    /** The toast to render */
    toast: QueuedToast<T>;
}

const Toast: React.FunctionComponent<ToastProps<ToastArgs>> = ({
    state,
    toast
}: ToastProps<ToastArgs>) => {
    const error = toast.content.status === "error";
    const duration = toast.timeout;
    const { message, isClosable } = toast.content;
    const toastRef = React.useRef<HTMLOutputElement>(null);
    const { toastProps, titleProps, closeButtonProps } = useToast(
        { toast },
        state,
        toastRef
    );

    delete closeButtonProps.onPress;

    React.useEffect(() => {
        if (duration !== undefined) {
            toastRef.current?.style.setProperty("--duration", `${duration}ms`);
        }
    }, [duration, toastRef]);

    const statusIcon = !error ? (
        <SuccessSolid className="ids-toast__icon" />
    ) : (
        <RemoveSolid className="ids-toast__icon" />
    );

    const classes = cx("ids-toast", {
        "ids-toast--error": error,
        "ids-toast--reduce-motion": duration === undefined
    });

    const toastElem = (
        <output
            {...toastProps}
            ref={toastRef}
            className={classes}
            data-animation={toast.animation}
            onAnimationEnd={() => {
                // Remove the toast when the exiting animation completes.
                if (toast.animation === "exiting") {
                    state.remove(toast.key);
                }
            }}
        >
            <div {...titleProps} className="ids-toast__title">
                {statusIcon} {message}
            </div>
            {isClosable && (
                <IconButton
                    {...closeButtonProps}
                    className="ids-toast__close"
                    onClick={() => {
                        state.close(toast.key);
                    }}
                    appearance={{
                        type: "ghost",
                        variant: "secondary"
                    }}
                    icon={<Close size="medium" />}
                />
            )}
            <div
                className="ids-toast__progress-indicator"
                onAnimationEnd={e => {
                    e.stopPropagation();
                }}
            />
        </output>
    );

    return toastElem;
};
export default Toast;
