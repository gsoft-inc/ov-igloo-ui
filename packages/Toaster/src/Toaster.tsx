import * as React from 'react';
import cx from 'classnames';

import SuccessSolid from '@igloo-ui/icons/dist/SuccessSolid';
import RemoveSolid from '@igloo-ui/icons/dist/RemoveSolid';
import Close from '@igloo-ui/icons/dist/Close';

import './toaster.scss';

export interface ToasterProps extends React.ComponentProps<'div'> {
  error?: boolean;
  children: React.ReactNode;
}

const Toaster: React.FunctionComponent<ToasterProps> = (
  props: ToasterProps
) => {
  const { children, error, className, ...rest } = props;

  const statusIcon = !error ? (
    <SuccessSolid className="ids-toaster__icon" />
  ) : (
    <RemoveSolid className="ids-toaster__icon" />
  );

  const classes = cx('ids-toaster', className, {
    'ids-toaster--error': error,
  });

  return (
    <div className={classes} {...rest}>
      <div className="ids-toaster__content">
        {statusIcon}
        <span className="ids-toaster__text">{children}</span>
      </div>
      <Close className="ids-toaster__close" />
    </div>
  );
};

export default Toaster;
