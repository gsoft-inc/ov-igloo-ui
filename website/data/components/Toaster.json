[
  [
    {
      "tags": {},
      "filePath": "/Users/franck.gaudin/Devel/Igloo/components/packages/Toaster/src/Toaster.tsx",
      "description": "",
      "displayName": "Toaster",
      "methods": [],
      "props": {
        "iconDescription": {
          "defaultValue": null,
          "description": "Add a aria-label to describe the close button",
          "name": "iconDescription",
          "parent": {
            "fileName": "components/packages/Toaster/src/Toaster.tsx",
            "name": "ToastProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Toaster/src/Toaster.tsx",
              "name": "ToastProps"
            }
          ],
          "required": false,
          "type": { "name": "\"string\"" }
        },
        "dataTest": {
          "defaultValue": null,
          "description": "Add a data-test tag for automated tests",
          "name": "dataTest",
          "parent": {
            "fileName": "components/packages/Toaster/src/Toaster.tsx",
            "name": "ToastProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Toaster/src/Toaster.tsx",
              "name": "ToastProps"
            }
          ],
          "required": false,
          "type": { "name": "string" }
        }
      }
    }
  ]
]
