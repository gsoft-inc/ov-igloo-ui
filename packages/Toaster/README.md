# Toaster

The Toaster component is a non-disruptive message that appears at the top of the viewport to provide quick, at-a-glance feedback on the outcome of an action.

<Example is="custom" />

<ReferenceLinks is="custom" />

## Installation

To install `@igloo-ui/toaster` in your project, you will need to run the following command using [npm](https://www.npmjs.com/):

```bash
npm install @igloo-ui/toaster
```

If you prefer [Yarn](https://classic.yarnpkg.com/en/), use the following command instead:

```bash
yarn add @igloo-ui/toaster
```

## Usage

### Displaying a toast

Then to use the component in your code just import it!

```jsx
import Toaster, { toast } from '@igloo-ui/toaster';
import Button from '@igloo-ui/button';

const App = () => {
  return (
    <div>
      <Button onClick={() => toast.success('Successfully toasted!')}>
        Success
      </Button>

      <Toaster />
    </div>
  );
};
```

### Displaying multiple toasts

Then to use the component in your code just import it!

```jsx
import Toaster, { toast } from '@igloo-ui/toaster';
import Button from '@igloo-ui/button';

const App = () => {
  return (
    <div>
      <Button onClick={() => toast.success('Successfully toasted!')}>
        Success
      </Button>
      <Button
        appearance="secondary"
        onClick={() => toast.error("This didn't work!")}
      >
        Error
      </Button>

      <Toaster />
    </div>
  );
};
```
