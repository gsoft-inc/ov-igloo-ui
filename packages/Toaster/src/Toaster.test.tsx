/**
 * @jest-environment jsdom
 */
import React from 'react';
import { shallow, mount } from 'enzyme';
import { Toast as HotToast } from 'react-hot-toast';

import Toaster, { toaster } from './Toaster';
import Toast from './Toast';
import { debug } from 'console';

const mockToast = {
  createdAt: 1639075944871,
  visible: true,
  ariaProps: {
    role: 'status',
    'aria-live': 'polite',
  },
  message: 'Successfully toasted!',
  pauseDuration: 0,
  position: 'top-center',
  duration: 5000,
  id: '1',
  style: {},
} as HotToast;

describe('Toaster', () => {
  const component = mount(
    <div>
      <button onClick={() => toaster.success('Successfully toasted!')}>
        Success
      </button>
      <Toaster />
    </div>
  );

  test('It should render without errors', () => {
    component.find('button').simulate('click');
    const wrapper = component.find('.ids-toaster');
    expect(wrapper.length).toBe(1);
  });

  test('It should render success', () => {
    expect(
      <Toast toast={{ ...mockToast, type: 'success' }} />
    ).toMatchSnapshot();
  });

  test('It should render error', () => {
    expect(<Toast toast={{ ...mockToast, type: 'error' }} />).toMatchSnapshot();
  });
});
