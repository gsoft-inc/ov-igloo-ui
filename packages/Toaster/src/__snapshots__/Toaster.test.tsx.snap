// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toaster It should render error 1`] = `
<Toast
  toast={
    Object {
      "ariaProps": Object {
        "aria-live": "polite",
        "role": "status",
      },
      "createdAt": 1639075944871,
      "duration": 5000,
      "id": "1",
      "message": "Successfully toasted!",
      "pauseDuration": 0,
      "position": "top-center",
      "style": Object {},
      "type": "error",
      "visible": true,
    }
  }
/>
`;

exports[`Toaster It should render success 1`] = `
<Toast
  toast={
    Object {
      "ariaProps": Object {
        "aria-live": "polite",
        "role": "status",
      },
      "createdAt": 1639075944871,
      "duration": 5000,
      "id": "1",
      "message": "Successfully toasted!",
      "pauseDuration": 0,
      "position": "top-center",
      "style": Object {},
      "type": "success",
      "visible": true,
    }
  }
/>
`;
